subsys/scsi/scsi.c

```diff
@@ -0,0 +1,513 @@
+	scsi_setup_write(&cmd, sector, count);
+
+	/* Execute the SCSI command and return the result */
+	ret = scsi_exec(scsi_dev, &cmd);
+
+	return ret;
+}
+
+/**
+ * @brief Perform a read operation on the SCSI device.
+ *
+ * This function sends a read command to the SCSI device.
+ *
+ * @param scsi_dev Pointer to the scsi_device structure.
+ * @param sector Starting sector number.
+ * @param count Number of sectors to read.
+ * @param buf Pointer to the buffer to store the read data.
+ *
+ * @return 0 if successful, or a negative error code on failure.
+ */
+int32_t scsi_read(struct scsi_device *scsi_dev, uint64_t sector,
+		  uint32_t count, uint8_t *buf)
+{
+	int32_t ret;
+	struct scsi_cmd cmd = {0};
+
+	if (scsi_dev == NULL) {
+		return -EINVAL;
+	}
+
+	/* Set up the SCSI command structure */
+	cmd.lun = scsi_dev->lun;
+	cmd.pdata = (uint8_t *)buf;
+	cmd.datalen = ((uint64_t)(scsi_dev->sector_size) * (uint64_t)(count));
+	cmd.dma_dir = (uint8_t)PERIPHERAL_TO_MEMORY;
+
+	/* Setup the "Read" command with specified sector and count */
+	scsi_setup_read(&cmd, sector, count);
+
+	/* Execute the SCSI command and return the result */
+	ret = scsi_exec(scsi_dev, &cmd);
+
+	return ret;
+}
+
+/**
+ * @brief Handle SG_IO ioctl for SCSI device.
+ *
+ * This function handles the SG_IO ioctl command for the SCSI device,
+ * which allows for sending arbitrary SCSI commands.
+ *
+ * @param sdev Pointer to the scsi_device structure.
+ * @param arg Pointer to the ioctl arguments.
+ *
+ * @return 0 if successful, or a negative error code on failure.
+ */
+static int32_t scsi_ioctl_sg_io(struct scsi_device *sdev, void *arg)
+{
+	const struct sg_io_req *req = (const struct sg_io_req *)arg;
+	struct scsi_cmd scmd = {0};
+	int32_t ret;
+
+	/* Validate the request */
+	if ((req == NULL) || (req->protocol != (uint32_t)BSG_PROTOCOL_SCSI) ||
+	    (req->subprotocol != (uint32_t)BSG_SUB_PROTOCOL_SCSI_CMD)) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* Initialize the SCSI command structure with ioctl request */
+	scmd.lun = sdev->lun;
+	scmd.cmdlen = (uint8_t)(req->request_len);
+	if (scmd.cmdlen > sizeof(scmd.cmd)) {
+		ret = -EINVAL;
+		goto out;
+	}
+	(void)memcpy((void *)&scmd.cmd[0], req->request, scmd.cmdlen);
+	scmd.dma_dir = (uint8_t)(req->dxfer_dir);
+	scmd.datalen = req->dxfer_len;
+	scmd.pdata = req->dxferp;
+
+	/* Execute the SCSI command and return the result */
+	ret = scsi_exec(sdev, &scmd);
+
+out:
+	return ret;
+}
+
+/**
+ * @brief Dispatch an ioctl to a SCSI device.
+ *
+ * This function dispatches the appropriate ioctl to the SCSI device based on
+ * the provided command.
+ *
+ * @param sdev Pointer to the scsi_device structure.
+ * @param cmd The ioctl command.
+ * @param arg Data associated with the ioctl.
+ *
+ * @return 0 if successful, or a negative error code on failure.
+ */
+int32_t scsi_ioctl(struct scsi_device *sdev, int32_t cmd, void *arg)
+{
+	int32_t ret = -EINVAL;
+
+	/* Validate that the device is not NULL */
+	if (sdev == NULL) {
```

Comment:
Generally no need for these checks, the application should check that a device is ready before attempting to use it